keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,172 @@
+Staging branch is automatically applied to the staging Kubernetes cluster by a push triggered Jenkins job.
+All members of her team file pull requests against the repo to bring changes into staging and production.
+
+Aside from that, Alice and her team got services that directly manage their objects, like controllers and webhooks,
+as well as services that updated labels for some objects.
+
+They want to make use of the new Server Side Apply, so concurrent changes to objects from different actors are possible without unwanted overlap/conflicts.
+
+To do so, assuming the changes in this KEP, Alice would add the newly introduced field to all manifests in their config management repo. The field would be set to e.g. `configManagement` (or their repo name).
+Additionally, she would update the manifests sent by different actors (like controllers) to contain the field with a value like `controllerName`. Note that this might already be the case for some through the defaulting to the user-agent inside the apply and update pathway.
+
+If now the team updates their repo with a change to their manifests, Jenkins again applies them as usual. On the apiserver side though, the fieldManager set in the applied manifest gets used and all fields set now get owned by the `configManagement` fieldManager.
+
+Again as usual other actors step in to change the objects. They as well use their fieldManager value either set in the manifest file they apply, or based on their user-agent/configuration.
+One of those actors is setting annotations for every service with external information.
+As the actor now only updates annotations not already owned by the `configManagement` fieldManager, applying them will cause the actor to own them.
+
+All actors now can coexist and manage their set of fields without interfering with each other.
+
+#### Kustomize
+
+In reference to the above workflow, it would be possible to override `options.fieldManager` for differen Kustomize overlays and therefor separate ownership depending on the actor currently active.
+
+This means, updating single fields by using a Kustomize overlay (like updating images, labels or annotations) as part of a CI/CD pipeline could easily contain the fieldManager information without any changes to Kustomize (or similar tools) itself.
+
+### Risks and Mitigations
+
+One risk might be, that introducing this change would mean there is a field that does not get persisted to storage.
+As a result, setting it causes different behavior on the apiserver, but is not reflected when reading the object back from the apiserver.
+
+This can seem unintuitive but should be solved by proper documentation.
+
+## Design Details
+
+### Proposed API Change
+
+```go
+type ObjectMeta struct {
+...
+  // Options used by the apiserver when handling the object.
+  // This field is write-only, non-persisted and optional.
+  Options *MetaOptions `json:"options,omitempty" protobuf:"bytes,18,opt,name=options"`
+...
+}
+
+type MetaOptions struct {
```

Comment:
is the eventual goal for this to contain more fields from updateOptions/createOptions etc?

what is the behavior if unrecognized options appear in this stanza? discard/ignore just like we do with unknown query parameters?

what is the behavior if conflicting options are specified in this stanza and the query? error? precedence order?

Reply:
The last question is addressed for this particular option in prose above, but I think it should be stated as a general principle in the comment on the Options field in the ObjectMeta struct.

At first I was going to say, I think the rule must be discard/ignore, which implies it will take multiple releases to roll out new options with this mechanism. But... this actually feels quite problematic. I think this should be addressed in the KEP in prose. The other option is e.g. custom encoding (via json.RawMessage or something) that permits us to detect and fail on unexpected entries. I think this would be important if we ever wanted to add a new field here. We actually may have this problem already given that we went to beta 1 without this.

I think we would choose to use this mechanism on a case-by-case basis; the only other current option is DryRun; I don't think it makes sense to put DryRun in a manifest but if someone can come up with a use case maybe we'd want to do it.

Reply:
DryRun could be used for automated testing but to set it this way would mean there is a manifest just for running DryRun so probably using the flag in tests is more likely.
Another possible field I could think of is force.

Reply:
This comment still seems to be pending: how do we safely add options to this, given that old apiservers won't understand them?

Reply:
I'll bring it up in our next wg meeting as I'm not completely sure about all the factors coming together here. Maybe @apelisse got an opinion on this.

Reply:
I feel like none of these problems would exist if we hadn't chosen the name "options", which I still really don't like. It sounds imperative, people writing manifests really don't know what it is, what it means and how it's used. It invites for more. The more I think about it, the more I suspect we should make it an annotation, and implement it in `cli-runtime`.

Reply:
> The more I think about it, the more I suspect we should make it an annotation, and implement it in `cli-runtime`.

Making clients deal with resources in a non-transparent manner doesn't seem like a good idea (increases cost across client libraries/languages, leads to inconsistency depending on the client, etc).

Reply:
Agreed, alternatively, I suggested "directives" rather than "options". I still really really don't like options. And we have possibly other directives that we may introduce so I think they would fit nicely there.

Reply:
+1 to @liggitt. 

Directives sounds non-declarative, which I still don't think these are? They are facts about the *manifest* rather than facts about the object once it is created. If they are facts about the API call only, then I'm less sure they belong here at all (I'd be even more strongly against putting them in annotations though).

Calling these "directives" doesn't solve the problem @liggitt brought up (of safely adding one when old apiservers don't understand them). I'm much more worried about this problem than the name.

Reply:
I agree that we should probably make this a `map[string]string` if we can be convinced that "string" is always good enough, and that we should reject unknown keys.